plugins/lua/buildingplan/unlink_mechanisms.lua

```diff
@@ -0,0 +1,682 @@
+                    {text = dfhack.items.getDescription(item, 0, true), pen = COLOR_BLUE},
+                },
+                desired_lines = 3,
+                propagate_fn = function(suppress)
+                    always_free_single = suppress
+                    item.flags.in_building = false
+                    dfhack.items.moveToGround(item, xyz2pos(dfhack.items.getPosition(item)))
+                end,
+            }:show()
+        end
+    else
+        dfhack.printerr(("MechItemOverlay: Invalid button! Offset %d"):format(button.frame.t))
+    end
+end
+
+function MechItemOverlay:ask_free_all()
+    if always_free_all then
+        self:do_free_all()
+        return
+    end
+
+    ConfirmScreen
+    {
+        title = "Free All Mechanisms",
+        message = "Really free all mechanisms? This cannot be undone.",
+        desired_lines = 1,
+        propagate_fn = function(suppress)
+            always_free_all = suppress
+            self:do_free_all()
+        end,
+    }:show()
+end
+
+function MechItemOverlay:do_free_all()
+    for item in mech_iter(self.building) do
+        if not get_mech_target(item) then
+            item.flags.in_building = false
+            dfhack.items.moveToGround(item, xyz2pos(dfhack.items.getPosition(item)))
+        end
+    end
+end
+
+function MechItemOverlay:update_buttons()
+    local scroll_pos = sheet.scroll_position_item
+    local bci_len = #self.building.contained_items
+    for i=1, self.num_buttons do
+        local button = self:get_button(i, true)
+        local offset = i*3 - 1 - ((scroll_pos + 1) % 3)
+        local idx = self:idx_from_offset(offset)
+
+        button.visible = false
+        if idx > 0 and idx < bci_len then
+            local item = self.building.contained_items[idx].item
+            if item._type == df.item_trappartsst and item.flags.in_building and
+                not item.flags.in_job and not get_trigger_index(item) then
+                    button.frame.t = offset
+                    button.visible = true
+            end
+        end
+        button:updateLayout()
+    end
+
+    self.subviews.free_all.visible = false
+    for item in mech_iter(self.building) do
+        if not get_mech_target(item) then
+            self.subviews.free_all.visible = true
+
+            local b = (self.frame.h % 3) == 1 and bci_len >= self.num_buttons and 0 or 1
+            self.subviews.free_all.frame.b = b --avoid overlapping list
+            self.subviews.free_all:updateLayout()
+            break
+        end
+    end
+end
+
+function MechItemOverlay:fix_layout()
+    for i=1, self.num_buttons do --hide existing buttons
+        local button = self:get_button(i)
+        if button then
+            button.visible = false
+        end
+    end
+
+    local h = self.parent_height - 46 - (has_link_tab(self.building) and 3 or 0)
+    self.frame.h = h + 1 --includes lower border
+    self.num_buttons = h // 3
+
+    self.frame.b = 3
+    self.frame.r = 40
```

Comment:
Probably shouldn't need to do this, but for some reason these both end up as 1 otherwise. Not sure why.